### PR TITLE
Ensure the stack is using a single ExchangeManager

### LINKED ISSUE #6390: [Crash] CHIP crashes sometimes when receiving a response from a CHIP accessory
This looks like a stack overflow since PeerAddress is not a pointer and `IsInitialized` should have just returned `false` if it wasn't init.

Note - this only happens when writing to the DoorLock Cluster. OnOff Cluster was working ok. Reports are enabled for that Cluster.

Last bit of Controller Logs:
```
2021-04-29 18:19:00.5785 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540594] [0x77511] CHIP: [ZCL] RX len 4, ep 1, clus 0x101
2021-04-29 18:19:00.5785 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540594] [0x77511] CHIP: [ZCL]  FC 9 seq 5 cmd 1 payload[
2021-04-29 18:19:00.5786 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540594] [0x77511] CHIP: [ZCL] 00
2021-04-29 18:19:00.5786 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540594] [0x77511] CHIP: [ZCL] ]
2021-04-29 18:19:00.5787 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540594] [0x77511] CHIP: [ZCL] UnlockDoorResponse:
2021-04-29 18:19:00.5787 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540594] [0x77511] CHIP: [ZCL]   status: EMBER_ZCL_STATUS_SUCCESS (0x00)
...
2021-04-29 18:19:00.6356 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540651] [0x77511] CHIP: [ZCL] RX len 13, ep 1, clus 0x101
2021-04-29 18:19:00.6357 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540651] [0x77511] CHIP: [ZCL]  FC 9 seq 12 cmd 20 payload[
2021-04-29 18:19:00.6357 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540651] [0x77511] CHIP: [ZCL] 01 02 00 00 00 00 00 00 00 00
2021-04-29 18:19:00.6358 -0700    info 0x77511 0x0 CHIPTool[21618] (CHIP): ðµ [1619745540651] [0x77511] CHIP: [ZCL] ]
```

Stack Trace:
```
Thread 4 Crashed:
0   CHIP                          	0x000000010155d450 chip::Transport::PeerAddress::IsInitialized() const + 12
1   CHIP                          	0x000000010155d3dc chip::Transport::PeerConnectionState::IsInitialized() + 36
2   CHIP                          	0x000000010155c134 chip::Transport::PeerConnections<16ul, (chip::Time::Source)0>::FindPeerConnectionState(chip::Optional<unsigned long long>, unsigned short, chip::Transport::PeerConnectionState*) + 132
3   CHIP                          	0x000000010155b8e0 chip::SecureSessionMgr::GetPeerConnectionState(chip::SecureSessionHandle) + 88
4   CHIP                          	0x000000010152d2e8 chip::Messaging::ExchangeContext::SendMessage(chip::Protocols::Id, unsigned char, chip::System::PacketBufferHandle, chip::BitFlags<chip::Messaging::SendMessageFlags, unsigned short> const&) + 124
5   CHIP                          	0x000000010142fb48 chipSendUnicast(unsigned long long, EmberApsFrame*, unsigned short, unsigned char*) + 660
6   CHIP                          	0x00000001014225f8 emAfSend(unsigned char, unsigned long long, EmberApsFrame*, unsigned char, unsigned char*, unsigned char*, unsigned short, unsigned char) + 232
7   CHIP                          	0x00000001014219bc send(unsigned char, unsigned long long, EmberApsFrame*, unsigned short, unsigned char*, bool, unsigned short, unsigned char, void (*)(unsigned char, unsigned long long, EmberApsFrame*, unsigned short, unsigned char*, unsigned char)) + 436
8   CHIP                          	0x0000000101421e14 emberAfSendUnicastWithCallback(unsigned char, unsigned long long, EmberApsFrame*, unsigned short, unsigned char*, void (*)(unsigned char, unsigned long long, EmberApsFrame*, unsigned short, unsigned char*, unsigned char)) + 164
9   CHIP                          	0x0000000101423698 emberAfSendResponseWithCallback(void (*)(unsigned char, unsigned long long, EmberApsFrame*, unsigned short, unsigned char*, unsigned char)) + 336
10  CHIP                          	0x0000000101423904 emberAfSendDefaultResponseWithCallback(EmberAfClusterCommand const*, EmberAfStatus, void (*)(unsigned char, unsigned long long, EmberApsFrame*, unsigned short, unsigned char*, unsigned char)) + 324
11  CHIP                          	0x0000000101423970 emberAfSendDefaultResponse(EmberAfClusterCommand const*, EmberAfStatus) + 40
12  CHIP                          	0x00000001014c2d80 emAfProcessClusterSpecificCommand(EmberAfClusterCommand*) + 208
13  CHIP                          	0x0000000101423354 dispatchZclMessage(EmberAfClusterCommand*) + 412
14  CHIP                          	0x0000000101422f58 emberAfProcessMessage(EmberApsFrame*, unsigned char, unsigned char*, unsigned short, unsigned long long, EmberAfInterpanHeader*) + 436
15  CHIP                          	0x00000001014c2ec0 HandleDataModelMessage(unsigned long long, chip::System::PacketBufferHandle) + 240
16  CHIP                          	0x00000001015086a0 chip::Controller::Device::OnMessageReceived(chip::PacketHeader const&, chip::PayloadHeader const&, chip::System::PacketBufferHandle) + 224
17  CHIP                          	0x000000010150b044 chip::Controller::DeviceController::OnMessageReceived(chip::Messaging::ExchangeContext*, chip::PacketHeader const&, chip::PayloadHeader const&, chip::System::PacketBufferHandle) + 292
18  CHIP                          	0x000000010152e1e4 chip::Messaging::ExchangeContext::HandleMessage(chip::PacketHeader const&, chip::PayloadHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) + 472
19  CHIP                          	0x000000010152feb8 chip::Messaging::ExchangeManager::OnMessageReceived(chip::PacketHeader const&, chip::PayloadHeader const&, chip::SecureSessionHandle, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle, chip::SecureSessionMgr*) + 996
20  CHIP                          	0x000000010155cd54 chip::SecureSessionMgr::SecureMessageDispatch(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) + 1144
21  CHIP                          	0x000000010155c83c chip::SecureSessionMgr::OnMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) + 120
22  CHIP                          	0x000000010155d8e4 chip::TransportMgrBase::HandleMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle) + 188
23  CHIP                          	0x00000001015ab4ac chip::Transport::Base::HandleMessageReceived(chip::PacketHeader const&, chip::Transport::PeerAddress const&, chip::System::PacketBufferHandle&&) + 156
24  CHIP                          	0x00000001015ab1f0 chip::Transport::UDP::OnUdpReceive(chip::Inet::IPEndPointBasis*, chip::System::PacketBufferHandle, chip::Inet::IPPacketInfo const*) + 144
25  CHIP                          	0x0000000101517744 chip::Inet::IPEndPointBasis::HandlePendingIO(unsigned short) + 1144
26  CHIP                          	0x000000010151e810 chip::Inet::UDPEndPoint::HandlePendingIO() + 88
27  CHIP                          	0x000000010151aa14 chip::Inet::InetLayer::HandleSelectResult(int, fd_set*, fd_set*, fd_set*) + 792
28  CHIP                          	0x00000001015429cc chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::SysProcess() + 300
29  CHIP                          	0x000000010154272c chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::_RunEventLoop() + 52
30  CHIP                          	0x0000000101542d58 chip::DeviceLayer::PlatformManager::RunEventLoop() + 28
31  CHIP                          	0x0000000101542b5c chip::DeviceLayer::Internal::GenericPlatformManagerImpl_POSIX<chip::DeviceLayer::PlatformManagerImpl>::EventLoopTaskMain(void*) + 52
32  libsystem_pthread.dylib       	0x00000001be26a8e4 0x1be269000 + 6372
33  libsystem_pthread.dylib       	0x00000001be273700 0x1be269000 + 42752
```

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem
 
`src/app/Server.cpp` is built by `src/lib`. As the result controllers ends up with 2 `chip::Messaging::ExchangeManager` living around.

The server code is not meant to be built for controllers.

 #### Summary of Changes
 * Remove the `src/app/server` dependency from `src/lib`
 * Add the dependency where it belongs
 * Move the (dirty) `Messaging::ExchangeManager & chip::ExchangeManager()` method a bit deeper in the stack so it can either returns the `ExchangeManager` created by `src/app/server/Server.cpp` or the one from the device controller.
 
 Fixes #6390

### CODE REVIEW

File: src/app/server/DataModelHandler.h [hunk 1]
```diff
@@ -30,7 +31,7 @@
  * data model messages.
  *
  */
-void InitDataModelHandler();
+void InitDataModelHandler(chip::Messaging::ExchangeManager * exchangeMgr);
```
Comment (bzbarsky-apple) [inline src/app/server/DataModelHandler.h:34]: I wonder whether this header should move (in a followup?) to src/app/util, which is where the .cpp lives....

File: src/app/server/Server.h [hunk 2]
```diff
@@ -66,7 +66,6 @@ enum class PairingWindowAdvertisement
 };
 
 SecureSessionMgr & SessionManager();
```
Comment (bzbarsky-apple) [inline src/app/server/Server.h:68]: Can we (followup?) remove this `SessionManager()` function and the extern decl in chip-message-send and the decl in `src/app/server/SessionManager.h` and the impl in `Server.cpp`?  No one uses it.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/server/DataModelHandler.h (hunk 1 above)
Comment (vivien-apple) [reply src/app/server/DataModelHandler.h:34]: > I wonder whether this header should move (in a followup?) to src/app/util, which is where the .cpp lives....

I agree. I notice that it lives in `src/app/server` too. Will fix that in a followup.

File: src/app/server/Server.h (hunk 2 above)
Comment (vivien-apple) [reply src/app/server/Server.h:68]: Done.

File: src/app/server/DataModelHandler.h (hunk 1 above)
Comment (vivien-apple) [reply src/app/server/DataModelHandler.h:34]: I end up fixing it here.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@msandstedt @andy31415 @saurabhst @hawk248 ?

**Review by saurabhst — APPROVED** [review]

**Comment by woody-apple** [comment]
@LuDuda @Damian-Nordic ?

**Review by andy31415 — APPROVED** [review]

**Pull request merged**